# RNA Architecture Underlies Discontinuous Transcription and Evolution of Coronavirus
RNA Architecture Underlies Discontinuous Transcription and Evolution of Coronavirus
Zi Wen1,2,3#, Lei Chen1,2,4#, Dehua Luo1,2,4#, Ju Sun1,2,4#, Liangrong Guo1,2,4, Yingxiang Deng1,2,4, Zhiyuan Huang2,3, Yuxiang Wang1,2,4, Ke Pan1,2,4, Fan Wang8, Shaobo Xiao1, Li Li2,3 and Dengguo Wei1,2,4,5,6,7

Correspondence: fwang@mail.ccnu.edu.cn; vet@mail.hzau.edu.cn; li.li@mail.hzau.edu.cn; dgwei@mail.hzau.edu.cn

Discontinuous transcription is employed by coronaviruses to generate subgenomic RNAs (sgRNAs) essential for gene expression and evolution adaptation. However, the mechanisms underlying the formation and functional roles of sgRNAs in specific coronavirus genomes are not well understood, particularly at the nucleotide level. We conducted an integrated analysis on 229 transcriptome and 12 RNA structurome samples across various coronavirus genera. RNA-RNA interactions between TRS-B and TRS-L flanking regions were identified in same genomic direction, which correlates with canonical junction formation. Non-canonical junctions frequently span either closely or distantly genomic distance, with short-range junctions generally mediated by stem-loops and overlapping with genomic deletion regions. Conserved long-range non-canonical sgRNAs were identified across different coronaviruses, and these sgRNAs harbor ORF10 or an evolving gene to suppress antiviral innate immune responses. This research highlights the importance of integrating transcriptome and RNA structurome profiles to elucidate the discontinuous transcription mechanism and evolution of coronaviruses.

Data can be  available on CODE (https://webofgroup.cn/dgwei/hngene).
