# RNA architecture underlies discontinuous transcription and evolution of coronavirus
RNA architecture underlies discontinuous transcription and evolution of coronavirus

Zi Wen, Lei Chen, Dehua Luo, Ju Sun, Zhiyuan Huang, Yingxiang Deng, Yuxiang Wang, Ke Pan, Liangrong Guo, Fan Wang*, Shaobo Xiao*, Li Li*, and Dengguo Wei*

* Correspondence: fwang@mail.ccnu.edu.cn; vet@mail.hzau.edu.cn; li.li@mail.hzau.edu.cn; dgwei@mail.hzau.edu.cn

Summary
Coronaviruses employ discontinuous transcription to generate subgenomic RNAs (sgRNAs) essential for encoding functional proteins, yet the precise mechanism underlying the sgRNAs generation remains elusive. To elucidate the spatial structural foundation of discontinuous transcription, we established the Coronavirus epigenOme DatabasE (CODE), a comprehensive repository integrating RNA architecture profiles and transcriptome data. We reveal that canonical junctions are facilitated by RNA-RNA interaction within same genomic direction, whereas non-canonical junctions are promoted by distal RNA-RNA interactions or local stem-loops. Moreover, we observe that short-range non-canonical junctions contribute to genomic deletions, and note long-range non-canonical sgRNAs harboring potential genes. Notably, formation of SL1, enriched with non-canonical junctions in SARS-CoV-2, is facilitated through genome recombination. These findings underscore the intricate interplay among RNA architecture, non-canonical junctions, and genetic mutations, suggesting a mutually adaptive relationship driving viral evolution. This study provides valuable insights into the mechanisms governing viral transcription and evolution at the RNA architecture level.

Data can be  available on CODE (https://webofgroup.cn/dgwei/hngene).
